posthog/hogql/transforms/property_types.py

```diff
@@ -0,0 +1,88 @@
+from typing import Dict, Set
+
+from posthog.hogql import ast
+from posthog.hogql.context import HogQLContext
+from posthog.hogql.parser import parse_expr
+from posthog.hogql.visitor import CloningVisitor, TraversingVisitor
+
+
+def resolve_property_types(node: ast.Expr, context: HogQLContext = None) -> ast.Expr:
+    from posthog.models import PropertyDefinition
+
+    # find all properties
+    property_finder = PropertyFinder()
+    property_finder.visit(node)
+
+    # fetch them
+    event_property_values = (
+        PropertyDefinition.objects.filter(
+            name__in=property_finder.event_properties,
+            team_id=context.team_id,
+            type__in=[None, PropertyDefinition.Type.EVENT],
+        ).values_list("name", "property_type")
+        if property_finder.event_properties
+        else []
+    )
+    event_properties = {name: property_type for name, property_type in event_property_values if property_type}
+
+    person_property_values = (
+        PropertyDefinition.objects.filter(
+            name__in=property_finder.person_properties,
+            team_id=context.team_id,
+            type=PropertyDefinition.Type.PERSON,
+        ).values_list("name", "property_type")
+        if property_finder.person_properties
+        else []
+    )
+    person_properties = {name: property_type for name, property_type in person_property_values if property_type}
```

Comment:
Might be quicker for me to ask than figure it out #Lazy

How often do we call this? 

Might be early-optimizing but the property types change infrequently so we can cache seen properties relatively aggressively. Makes me wonder how well our tracing covers this code 🤔

Reply:
This will just be called once per query in most cases. In case of non-hogql insights, it'll be called once per filter. This is where caching could make the biggest difference, but I didn't feel like overoptimising yet.